document/execution/instructions.rst

```diff
@@ -0,0 +1,1053 @@
+
+.. _exec-instr-seq-exit:
+
+Exiting :math:`\instr^\ast` with label :math:`L`
+................................................
+
+When the end of a labelled instruction sequence is reached without a jump or trap aborting it, then the following steps are performed.
+
+1. Let :math:`n` be the arity of :math:`L`.
+
+2. Assert: due to :ref:`validation <valid-instr-seq>`, there are :math:`n` values on the top of the stack.
+
+3. Pop the results :math:`\val^n` from the stack.
+
+4. Assert: due to :ref:`validation <valid-instr-seq>`, the label :math:`L` is now on the top of the stack.
+
+5. Pop the label from the stack.
+
+6. Push :math:`\val^n` back to the stack.
+
+7. Jump to the position after the |END| of the :ref:`structured control instruction <syntax-instr-control>` associated with the label :math:`L`.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   \LABEL_n\{\instr^\ast\}~\val^n~\END &\stepto& \val^n
+   \end{array}
+
+.. note::
+   This semantics also applies to the instruction sequence contained in a |LOOP| instruction.
+   Therefor, execution of a loop falls off the end, unless a backwards branch is performed explicitly.
+
+
+.. index:: ! invocation, function, function instance, label, frame
+
+Function Calls
+~~~~~~~~~~~~~~
+
+.. _exec-invoke:
+
+Invocation of :ref:`function address <syntax-funcaddr>` :math:`a`
+.................................................................
+
+1. Assert: due to :ref:`validation <valid-call>`, :math:`S.\FUNCS[a]` exists.
+
+2. Let :math:`f` be the :ref:`function instance <sytnax-funcinst>`, :math:`S.\FUNCS[a]`.
+
+3. Assert: due to :ref:`validation <valid-func>`, :math:`f.\MODULE.\TYPES[f.\CODE.\TYPE]` exists.
+
+4. Let :math:`[t_1^n] \to [t_2^m]` be the :ref:`function type <syntax-functype>` :math:`f.\MODULE.\TYPES[f.\CODE.\TYPE]`.
+
+5. Let :math:`t^\ast` be the list of :ref:`value types <syntax-valtype>` :math:`f.\CODE.\LOCALS`.
+
+6. Let :math:`\instr^\ast~\END` be the :ref:`expression <syntax-expr>` :math:`f.\CODE.\BODY`.
+
+7. Assert: due to :ref:`validation <valid-call>`, :math:`n` values are on the top of the stack.
+
+8. Pop the values :math:`\val^n` from the stack.
+
+9. Let :math:`\val_0^\ast` be the list of zero values of types :math:`t^\ast`.
+
+10. Let :math:`F` be the :ref:`frame <syntax-frame>` :math:`\{ \MODULE~f.\MODULE, \LOCALS~\val^n~\val_0^\ast \}`.
+
+11. Push the activation of :math:`F` with arity :math:`m` to the stack.
+
+12. :ref:`Execute <exec-block>` the instruction :math:`\BLOCK~[t_2^m]~\instr^\ast~\END`.
+
+.. math::
+   \begin{array}{l}
+   \begin{array}{lcl@{\qquad}l}
+   \val^n~(\INVOKE~a) &\stepto& \FRAME_m\{F\}~\BLOCK~[t_2^m]~\instr^\ast~\END~\END
+   \end{array}
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\mbox{if} & S.\FUNCS[a] = f \\
+     \wedge & f.\CODE = \{ \TYPE~x, \LOCALS~t^\ast, \BODY~\instr^\ast~\END \} \\
+     \wedge & f.\MODULE.\TYPES[x] = [t_1^n] \to [t_2^m] \\
+     \wedge & F = \{ \MODULE~f.\MODULE, ~\LOCALS~\val^n~(t.\CONST~0)^\ast \})
```

Comment:
Ah, I see. The mapping normally tracks for me, but `t* -> (t.const 0)*` didn't as much. In my mind using another symbol would make it clearer (e.g. `t^k -> (t.const 0)^k`), but I'm no formalist so I don't know if that makes any sense. :-)